pkg/virt-controller/watch/vm.go

```diff
@@ -1341,6 +1379,114 @@ func (c *VMController) updateStatus(vmOrig *virtv1.VirtualMachine, vmi *virtv1.V
 	return err
 }
 
+func (c *VMController) setPrintableStatus(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) {
+	// For each status, there's a separate function that evaluates
+	// whether the status is "true" for the given VM.
+	//
+	// Note that these statuses aren't mutually exclusive,
+	// and several of them can be "true" at the same time
+	// (e.g., Running && Migrating, or Paused && Terminating).
+	//
+	// The actual precedence of these statuses are determined by the order
+	// of evaluation - first match wins.
+	statuses := []struct {
+		statusType virtv1.VirtualMachinePrintableStatus
+		statusFunc func(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) bool
+	}{
+		{virtv1.VirtualMachineStatusTerminating, c.isVirtualMachineStatusTerminating},
+		{virtv1.VirtualMachineStatusStopping, c.isVirtualMachineStatusStopping},
+		{virtv1.VirtualMachineStatusMigrating, c.isVirtualMachineStatusMigrating},
+		{virtv1.VirtualMachineStatusStopped, c.isVirtualMachineStatusStopped},
+		{virtv1.VirtualMachineStatusProvisioning, c.isVirtualMachineStatusProvisioning},
+		{virtv1.VirtualMachineStatusStarting, c.isVirtualMachineStatusStarting},
+		{virtv1.VirtualMachineStatusPaused, c.isVirtualMachineStatusPaused},
+		{virtv1.VirtualMachineStatusRunning, c.isVirtualMachineStatusRunning},
+	}
+
+	for _, status := range statuses {
+		if status.statusFunc(vm, vmi) {
+			vm.Status.PrintableStatus = status.statusType
+			return
```

Comment:
I put the Q here - but it's a general one - Are you sure the status conditions are all mutual exclusive ?

Reply:
No, see the code comment just above that:
```go
// For each status, there's a separate function that evaluates
// whether the status is "true" for the given VM.
//
// Note that these statuses aren't mutually exclusive,
// and several of them can be "true" at the same time
// (e.g., Running && Migrating, or Paused && Terminating).
//
// The actual precedence of these statuses are determined by the order
// of evaluation - first match wins.
```

Reply:
This seems problematic isn't it ?  So in theory someone can run the `get` twice on a VM while nothing changed but still get two different status outcomes ?

Reply:
The above comment relates to these "status funcs" being not mutually exclusive. It means that two or more of these functions can (potentially) evaluate to true at the same time, and hence their order of evaluation is important.